File: src/infinicore/ops/syrk/syrk.cc
#include "infinicore/ops/syrk.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Syrk);

Syrk::Syrk(const Tensor &a, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, alpha, beta, c);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), a, alpha, beta, c, uplo, trans);
}

void Syrk::execute(const Tensor &a, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Syrk, a, alpha, beta, c, uplo, trans);
}

void syrk_(const Tensor &a, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    Syrk::execute(a, alpha, beta, c, uplo, trans);
}

} // namespace infinicore::op

File: python/infinicore/ops/syrk.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def syrk(
    a: Tensor,
    alpha: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    uplo: int = 0,
    trans: int = 0,
):
    _infinicore.syrk_(
        a._underlying,
        alpha._underlying,
        beta._underlying,
        out._underlying,
        uplo,
        trans,
    )

    return out
